### PR TITLE
core-edge: refactor log entry supplier

### PR DESCRIPTION
**Pull request by martinfurmanski**

### CODE REVIEW

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 1]
```diff
@@ -134,57 +134,34 @@ public synchronized void notifyCommitted( long commitIndex )
...
     {
         applier.submit( ( status ) -> () -> {
-            try ( LogEntrySupplier cacheLogEntrySupplier = new CacheApplier();
-                    LogEntrySupplier cursorLogEntrySupplier = new CursorApplier() )
+            try ( LogEntrySupplier logEntrySupplier = new LogEntrySupplier() )
             {
-                LogEntrySupplier currentLogEntrySupplier = cacheLogEntrySupplier;
                 lastApplyingStorage.persistStoreData( lastToApply );
-                for ( long i = lastApplied + 1; !status.isCancelled() &&  i <= lastToApply; i++ )
+                for ( long logIndex = lastApplied + 1; !status.isCancelled() &&  logIndex <= lastToApply; logIndex++ )
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: Imo it is better to move `!status.isCancelled()` inside `for` as an additional `if` statement. It would be more readable.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 2]
```diff
@@ -196,40 +173,41 @@ private void submitApplyJob( long lastToApply )
         } );
     }
 
-    private interface LogEntrySupplier extends AutoCloseable
+    private class LogEntrySupplier implements AutoCloseable
     {
-        RaftLogEntry get( long indexToApply ) throws IOException;
-    }
+        private RaftLogCursor cursor = null;
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: Why this explicit initialization is needed?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 1 above)
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: OK but it's not part of this PR.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 2 above)
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: It isn't, it is the default value in Java.

Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: Let's than trust Java and not do explicit initialization with default values :)

Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: They are equivalent really

Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: You mean `null` that we assign and that Java assigns?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 3]
```diff
@@ -134,57 +134,34 @@ public synchronized void notifyCommitted( long commitIndex )
...
-                        handleBatch( i - 1 /* ignore the current entry */, distributedOperations );
-                        assertIndexIsValidAndIncrementLastApplied( i, 1 );
+                        handleBatch( logIndex - 1 /* ignore the current entry */, distributedOperations );
+                        assertIndexIsValidAndIncrementLastApplied( logIndex, 1 );
                     }
-
-                    // no matter what, we need to purge the inflight cache from applied entries
-                    inFlightMap.unregister( i );
                 }
                 handleBatch( lastToApply, distributedOperations );
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: The introduced break instead of throwing an Exception makes lastToApply the wrong value. Needs a test for the behaviour in light of truncation.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 4]
```diff
@@ -134,57 +134,34 @@ public synchronized void notifyCommitted( long commitIndex )
...
-                            handleBatch( i, distributedOperations );
+                            handleBatch( logIndex, distributedOperations );
                         }
                     }
                     else
                     {
-                        handleBatch( i - 1 /* ignore the current entry */, distributedOperations );
-                        assertIndexIsValidAndIncrementLastApplied( i, 1 );
+                        handleBatch( logIndex - 1 /* ignore the current entry */, distributedOperations );
+                        assertIndexIsValidAndIncrementLastApplied( logIndex, 1 );
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: Should this increment really only happen in this else-branch? Looks weird to me.

Not sure when introduced, but this is wrong right? @davidegrohmann

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 1 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: I disagree, having the condition in one place is a more readable solution than having `continue` or `break` in the `for` loop.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 4 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: No it is correct, you need to increment one more time to "skip" the entry that doesn't contain a `DistributedOperation`.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 3 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: I think you should throw and panic in case you cannot read the entry. It is an unexpected behaviour that in theory should never happen.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 2 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: can we please remove the `null` initialization and stop this silly discussion?

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 3 above)
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: +1 for test

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 4 above)
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: @davidegrohmann Yea so what I meant was the assertIndexIsValidAndIncrementLastApplied, which only happens in the else-branch. But there seems to also be a call to that inside handleBatch. I am refactoring this anyway, so not important.

Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: fair enough

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 3 above)
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: +1 for should never happen, my mistake, there was already a test for it
